amazon-kinesis-client/src/main/java/software/amazon/kinesis/lifecycle/ShutdownTask.java

```diff
@@ -203,6 +209,38 @@ private void updateLeasesForChildShards()
         log.info("Shard {}: Updated current lease {} with child shard information: {}", shardInfo.shardId(), currentLease.leaseKey(), childShardIds);
     }
 
+    private void garbageCollectLeaseForParentShard() throws DependencyException, ProvisionedThroughputException, InvalidStateException {
+       final Lease parentShardLease = leaseCoordinator.getCurrentlyHeldLease(leaseKeyProvider.apply(shardInfo));
+
+        if (CollectionUtils.isNullOrEmpty(parentShardLease.childShardIds())) {
+            log.warn("No child shard state found in lease {} - not deleting lease.", parentShardLease);
+        } else {
+            final Set<String> allChildShardLeases = parentShardLease.childShardIds().stream()
+                    .map(c -> ShardInfo.getLeaseKey(shardInfo, c)).collect(Collectors.toSet());
+            final Set<String> processedChildShardLeases = new HashSet<>();
+
+            for (String childShard : parentShardLease.childShardIds()) {
+                final String leaseKey = ShardInfo.getLeaseKey(shardInfo, childShard);
+                final Lease lease = leaseCoordinator.leaseRefresher().getLease(leaseKey);
+
+                // We should only be deleting the parent shard lease if all of its children are currently being
+                // processed, i.e their checkpoint is not TRIM_HORIZON.
+                if (lease != null && !lease.checkpoint().equals(ExtendedSequenceNumber.TRIM_HORIZON)) {
+                    processedChildShardLeases.add(leaseCoordinator.leaseRefresher().getLease(leaseKey).leaseKey());
+                }
+            }
+
+            if (Objects.equals(allChildShardLeases, processedChildShardLeases)) {
+                log.info("Deleting lease {} as it has been completely processed and processing of child shards has begun.",
+                        parentShardLease);
+                leaseCoordinator.leaseRefresher().deleteLease(parentShardLease);
```

Comment:
Shouldn't we also check if the parent lease of this lease is deleted before we delete this lease?

Reply:
Added